### PR TITLE
RUN: Support build tool window

### PR DESCRIPTION
**Pull request by mchernyavsky**
Closes https://github.com/intellij-rust/intellij-rust/issues/2638. Fixes https://github.com/intellij-rust/intellij-rust/issues/3777. Fixes https://github.com/intellij-rust/intellij-rust/issues/3843.

toolwindow

Overview:
- TODO

https://github.com/rust-lang/cargo/issues/6994

### CODE REVIEW

File: src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
+        lineBuffer.add(event.text)
+        if (event.text.endsWith('\n')) {
+            // progress message [K actual message
+            val lines = lineBuffer.joinToString("").split("[K")
+            lineBuffer.clear()
+
+            // don't print progress message
+            buildOutputParser.parseOutput(lines.last(), outputType == ProcessOutputTypes.STDOUT) {
```
Comment (ortem) [inline src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt]: Can `lines` be empty (`NoSuchElementException`)?

Comment (mchernyavsky) [reply src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt]: No, see line 80.

**Comment by lancelote** [comment]
Should it report syntax errors as well? E.g. running something stupid like

```
fn main() {
    foo(
}
```

results in "Build failed" without much details so I have to run `cargo` in the terminal to see the error message:

Screen Shot 2019-08-01 at 14 14 43

**Comment by mchernyavsky** [comment]
@lancelote Thanks! Should be fixed now.

**Comment by mchernyavsky** [comment]
bors try

File: src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    override fun startNotified(event: ProcessEvent) {
+        val descriptor = DefaultBuildDescriptor(
+            context.buildId,
+            "Run Cargo command",
+            context.workingDirectory.toString(),
+            context.started
+        )
+        val buildStarted = StartBuildEventImpl(descriptor, "${context.taskName} running...")
+            .withExecutionFilters(*createFilters(context.cargoProject).toTypedArray())
+            .withRestartAction(createStopAction(event.processHandler))
```
[the remark this answers is not in the document]
Comment (ortem) [reply src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt]: @Undin Could you please take a look at this action and the corresponding code in `com.intellij.build.BuildView#createConsoleActions`?

**Comment by Undin** [comment]
> The build toolwindow ignores --quite option. It should also ignore -q option

Why?

**Comment by mchernyavsky** [comment]
@Undin With `--quite` option Cargo still shows compiler messages (errors & warnings), but we need Cargo messages to correctly display errors in the build toolwindow.

**Comment by lancelote** [comment]
@mchernyavsky After disabling `cargo.build.tool.window.enabled` I can no longer run `build` run configurations. Clicking buttons has no effect, no relevant logs are recorded as well.

**Pull request merged**